### PR TITLE
feat: add action for mc undo recursive

### PR DESCRIPTION
**Pull request by jiuker**
## Community Contribution License
All community contributions in this pull request are licensed to the project maintainers
under the terms of the [Apache 2 license](https://www.apache.org/licenses/LICENSE-2.0). 
By creating this pull request I represent that I have the right to license the 
contributions to the project maintainers under the Apache 2 license.

## Description

`mc undo --recursive minio/mytest/bin —force —action DELETE`
Support the top version is DELETE will do that or do nothing.
`mc undo --recursive minio/mytest/bin —force —action PUT`
Support the top version is PUT will do that or do nothing.

## Types of changes
- [x] Bug fix [[(non-breaking change which fixes an issue)]()](https://github.com/minio/mc/issues/4693)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fmc%2fpull%2fNNNNN)

### CODE REVIEW

File: cmd/undo-main.go [hunk 1]
```diff
@@ -218,17 +223,23 @@ func undoURL(ctx context.Context, aliasedURL string, last int, recursive, dryRun
...
-			exitErr = undoLastNOperations(ctx, clnt, perObjectVersions, last, dryRun)
+			if remove {
+				exitErr = undoLastNOperations(ctx, clnt, perObjectVersions, last, dryRun)
+			}
+			remove = true
 			lastObjectPath = content.URL.Path
 			perObjectVersions = []*ClientContent{}
 		}
-
+		if content.IsLatest && action == "DELETE" && !content.IsDeleteMarker || content.IsLatest && action == "PUT" && content.IsDeleteMarker {
```
Comment (shtripat) [inline cmd/undo-main.go:233]: Do you want to club conditions as `((content.IsLatest && action == "DELETE" && !content.IsDeleteMarker) || (content.IsLatest && action == "PUT" && content.IsDeleteMarker))` for better readability?

**Review by shtripat — COMMENTED** [review]
Can you please explain the use case why we need this in PR description?

**Comment by jiuker** [comment]
> Can you please explain the use case why we need this in PR description?

The issue https://github.com/minio/mc/issues/4693 need this feature.

File: cmd/undo-main.go [hunk 2]
```diff
@@ -218,17 +223,23 @@ func undoURL(ctx context.Context, aliasedURL string, last int, recursive, dryRun
...
-			exitErr = undoLastNOperations(ctx, clnt, perObjectVersions, last, dryRun)
+			if remove {
+				exitErr = undoLastNOperations(ctx, clnt, perObjectVersions, last, dryRun)
+			}
+			remove = true
 			lastObjectPath = content.URL.Path
 			perObjectVersions = []*ClientContent{}
 		}
-
+		if (content.IsLatest && action == "DELETE" && !content.IsDeleteMarker) || (content.IsLatest && action == "PUT" && content.IsDeleteMarker) {
```
Comment (klauspost) [inline cmd/undo-main.go:233]:
```suggestion
```suggestion
		if !remove {
			continue
		}
		if (content.IsLatest && action == "DELETE" && !content.IsDeleteMarker) || (content.IsLatest && action == "PUT" && content.IsDeleteMarker) {
```

**Review by klauspost — CHANGES_REQUESTED** [review]

File: cmd/undo-main.go (hunk 2 above)
Comment (jiuker) [reply cmd/undo-main.go:238]: @klauspost This is too powerful!

File: cmd/undo-main.go [hunk 3]
```diff
@@ -127,6 +131,10 @@ func parseUndoSyntax(ctx *cli.Context) (targetAliasedURL string, last int, recur
 	}
 
 	dryRun = ctx.Bool("dry-run")
+	action = strings.ToUpper(ctx.String("action"))
+	if action != "PUT" && action != "DELETE" && action != "" {
+		fatalIf(errInvalidArgument().Trace(), "Invalid action, should be PUT,DELETE or empty")
```
Comment (shtripat) [inline cmd/undo-main.go:136]:
```suggestion
		fatalIf(errInvalidArgument().Trace(), "Invalid action, should be PUT, DELETE or empty")
```

**Review by shtripat — APPROVED** [review]
Looks good. A cosmetic one.

**Review by harshavardhana — APPROVED** [review]

File: cmd/undo-main.go [hunk 4]
```diff
@@ -49,6 +49,10 @@ var undoFlags = []cli.Flag{
 		Name:  "dry-run",
 		Usage: "fake an undo operation",
 	},
+	cli.StringFlag{
+		Name:  "action",
```
Comment (vadmeste) [inline cmd/undo-main.go:53]: Unfortunately --action and --last doesn't work well together, maybe we should require last=1 if --action is specified.

**Review by vadmeste — CHANGES_REQUESTED** [review]

File: cmd/undo-main.go (hunk 4 above)
Comment (jiuker) [reply cmd/undo-main.go:53]: Ok

File: cmd/undo-main.go [hunk 5]
```diff
@@ -127,6 +131,13 @@ func parseUndoSyntax(ctx *cli.Context) (targetAliasedURL string, last int, recur
 	}
 
 	dryRun = ctx.Bool("dry-run")
+	action = strings.ToUpper(ctx.String("action"))
+	if action != "PUT" && action != "DELETE" && action != "" {
+		fatalIf(errInvalidArgument().Trace(), "Invalid action, should be PUT, DELETE or empty")
```
Comment (harshavardhana) [inline cmd/undo-main.go:136]:
```suggestion
		fatalIf(errInvalidArgument().Trace(), "unsupported action specified, supported actions are PUT, DELETE or empty (default)")
```

File: cmd/undo-main.go [hunk 6]
```diff
@@ -49,6 +49,10 @@ var undoFlags = []cli.Flag{
 		Name:  "dry-run",
 		Usage: "fake an undo operation",
 	},
+	cli.StringFlag{
+		Name:  "action",
+		Usage: "undo only if latest version is this type [PUT/DELETE]",
```
Comment (harshavardhana) [inline cmd/undo-main.go:54]:
```suggestion
		Usage: "undo only if the latest version is of the following type [PUT/DELETE]",
```

File: cmd/undo-main.go [hunk 7]
```diff
@@ -127,6 +131,13 @@ func parseUndoSyntax(ctx *cli.Context) (targetAliasedURL string, last int, recur
 	}
 
 	dryRun = ctx.Bool("dry-run")
+	action = strings.ToUpper(ctx.String("action"))
+	if action != "PUT" && action != "DELETE" && action != "" {
+		fatalIf(errInvalidArgument().Trace(), "Invalid action, should be PUT, DELETE or empty")
+	}
+	if (action == "PUT" || action == "DELETE") && last != 1 {
+		fatalIf(errInvalidArgument().Trace(), "This is a dangerous operation, you need to provide --last=1")
```
Comment (harshavardhana) [inline cmd/undo-main.go:139]:
```suggestion
		fatalIf(errInvalidArgument().Trace(), "--action if specified requires that you must specify --last=1")
```

File: cmd/undo-main.go [hunk 8]
```diff
@@ -215,20 +226,28 @@ func undoURL(ctx context.Context, aliasedURL string, last int, recursive, dryRun
...
+			}
+			remove = true
 			lastObjectPath = content.URL.Path
 			perObjectVersions = []*ClientContent{}
 		}
-
+		if !remove {
+			continue
+		}
+		if (content.IsLatest && action == "DELETE" && !content.IsDeleteMarker) || (content.IsLatest && action == "PUT" && content.IsDeleteMarker) {
```
Comment (harshavardhana) [inline cmd/undo-main.go:241]:
```suggestion
		if (content.IsLatest && action == "DELETE" && !content.IsDeleteMarker) || (content.IsLatest && action == "PUT" && content.IsDeleteMarker) {
```

Convert actions into constants such as 

```
const (
  actionPut    = "PUT"
  actionDelete = "DELETE"
)
```

It is generally preferred to have constants for strings like these.   actionPut = "PUT"

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Review by harshavardhana — APPROVED** [review]

**Review by shtripat — APPROVED** [review]

**Pull request merged**